python/ray/dag/tests/experimental/test_execution_schedule.py

```diff
@@ -0,0 +1,1017 @@
+    graph_nodes = {}
+    for op_type in DAGNodeOperationType:
+        graph_nodes[op_type] = DAGOperationGraphNode(
+            DAGNodeOperation(local_idx, op_type),
+            global_idx,
+            actor_handle,
+            requires_nccl,
+        )
+    return graph_nodes
+
+
+class TestSelectNextNodes:
+    """
+    Test whether `_select_next_nodes` function selects the next nodes for
+    topological sort to generate execution schedule correctly.
+
+    global_idx: Each DAG node has a unique global index.
+    local_idx: The DAG node's index in the actor's `executable_tasks` list.
+    """
+
+    def test_two_candidates_on_same_actor(self):
+        """
+        Simulate the case where there are two candidates on the same actor.
+        The candidate with the smaller index in the `executable_tasks` list
+        should be selected.
+
+        driver -> fake_actor.op -> fake_actor.op -> driver
+
+        In the example above, both READ operations on the fake_actor have zero
+        in-degree. The operation with the smaller index in the executable_tasks
+        list should be selected first; therefore, the one on the left side will
+        be selected first.
+        """
+        fake_actor = "fake_actor"
+        # The DAG node has a global index of 1, and its index in the
+        # actor's `executable_tasks` list is 0.
+        global_idx_1 = 1
+        dag_node_1 = DAGOperationGraphNode(
+            DAGNodeOperation(0, DAGNodeOperationType.READ),
+            global_idx_1,
+            fake_actor,
+            False,
+        )
+        # The DAG node has a global index of 2, and its index in the
+        # actor's `executable_tasks` list is 1.
+        global_idx_2 = 2
+        dag_node_2 = DAGOperationGraphNode(
+            DAGNodeOperation(1, DAGNodeOperationType.READ),
+            global_idx_2,
+            fake_actor,
+            False,
+        )
+        mock_actor_to_candidates = {
+            fake_actor: [
+                dag_node_1,
+                dag_node_2,
+            ],
+        }
+        next_nodes = _select_next_nodes(mock_actor_to_candidates, None)
+        assert len(next_nodes) == 1
+        assert next_nodes[0] == dag_node_1
+
+    def test_only_one_nccl_write(self, monkeypatch):
+        """
+        Simulate the case where there is only one candidate which is a NCCL
+        WRITE operation. In this case, `_select_next_nodes` should return both
+        the NCCL WRITE operation and the corresponding READ operation.
+
+        driver -> fake_actor_1.op -> fake_actor_2.op -> driver
+
+        In the example above, communication between fake_actor_1 and fake_actor_2
+        is done using NCCL. The following test case simulates a scenario where the
+        READ and COMPUTE operations on fake_actor_1 have already been added to the
+        execution schedule.
+        """
+        monkeypatch.setattr(ActorHandle, "__init__", mock_actor_handle_init)
```

Comment:
Why do we need to mock this?

Reply:
`_select_next_nodes` needs the information of `actor_handle._actor_id` because the key of `actor_to_candidates` is `ray._raylet.ActorID`.